Src/VimWpf/Implementation/BlockCaret/BlockCaret.cs

```diff
@@ -424,55 +480,54 @@ private Size CalculateCaretSize(out string caretCharacter)
+                        width = textViewLine.GetCharacterBounds(point).Width;
                     }
                     else if (codePointInfo == CodePointInfo.SurrogatePairHighCharacter)
                     {
                         // Handle surrogate pairs.
                         caretCharacter = new SnapshotSpan(point, 2).GetText();
-                        width = line.GetCharacterBounds(point).Width;
+                        width = textViewLine.GetCharacterBounds(point).Width;
                     }
                     else if (pointCharacter == '\t')
                     {
                         // Handle tab as no character and default width,
                         // except no wider than the tab's screen width.
                         caretCharacter = "";
-                        width = Math.Min(defaultWidth, line.GetCharacterBounds(point).Width);
+                        width = Math.Min(defaultWidth, textViewLine.GetCharacterBounds(point).Width);
                     }
                     else if (pointCharacter == '\r' || pointCharacter == '\n')
                     {
                         // Handle linebreak.
                         caretCharacter = "";
-                        width = line.GetCharacterBounds(point).Width;
+                        width = textViewLine.GetCharacterBounds(point).Width;
                     }
                     else
                     {
                         // Handle ordinary UTF16 character.
                         caretCharacter = pointCharacter.ToString();
-                        width = line.GetCharacterBounds(point).Width;
+                        width = textViewLine.GetCharacterBounds(point).Width;
                     }
                 }
             }
 
             return new Size(width, height);
         }
 
-        private double CalculateBaselineOffset()
+        private double CalculateBaselineOffset(VirtualSnapshotPoint caretPoint)
         {
             var offset = 0.0;
-            if (IsRealCaretVisible)
+            if (IsRealCaretVisible(caretPoint, out var textViewLine))
             {
-                var line = TextViewLineContainingCaret;
-                offset = Math.Max(0.0, line.Baseline - _formattedText.Baseline);
+                offset = Math.Max(0.0, textViewLine.Baseline - _formattedText.Baseline);
             }
             return offset;
         }
 
-        private Tuple<Rect, double, string> CalculateCaretRectAndDisplayOffset()
+        private Tuple<Rect, double, string> CalculateCaretRectAndDisplayOffset(VirtualSnapshotPoint caretPoint)
         {
-            var size = CalculateCaretSize(out string caretCharacter);
-            var caretPoint = GetRealCaretVisualPoint();
-            var blockPoint = caretPoint;
+            var size = CalculateCaretSize(caretPoint, out string caretCharacter);
+            var point = GetRealCaretVisualPoint(caretPoint);
+            var blockPoint = point;

```

Comment:
Is it okay for this logic to run when `point` has the vale `s_invalidPoint`?

Reply:
I will check this.